File: packages/contracts/src/canonical-json.ts
import {
  evaluate,
  parse,
  type JSONValue,
  type ObjectNode,
  type ValueNode,
} from "@humanwhocodes/momoa";
import canonicalize from "canonicalize";

export type ContractReason =
  | "canonical"
  | "digest"
  | "duplicate_key"
  | "enum"
  | "format"
  | "length"
  | "number"
  | "range"
  | "required"
  | "schema_version"
  | "stream"
  | "type"
  | "unicode"
  | "unknown_field"
  | "unsafe_path";

export class ContractInputError extends Error {
  readonly pointer: string;
  readonly reason: ContractReason;

  constructor(reason: ContractReason, pointer: string, message: string) {
    super(message);
    this.name = "ContractInputError";
    this.pointer = pointer;
    this.reason = reason;
  }
}

function escapePointer(value: string): string {
  return value.replaceAll("~", "~0").replaceAll("/", "~1");
}

function childPointer(parent: string, child: string | number): string {
  return `${parent}/${typeof child === "number" ? String(child) : escapePointer(child)}`;
}

function hasUnpairedSurrogate(value: string): boolean {
  for (let index = 0; index < value.length; index += 1) {
    const code = value.charCodeAt(index);
    if (code >= 0xd800 && code <= 0xdbff) {
      const next = value.charCodeAt(index + 1);
      if (next < 0xdc00 || next > 0xdfff) {
        return true;
      }
      index += 1;
    } else if (code >= 0xdc00 && code <= 0xdfff) {
      return true;
    }
  }
  return false;
}

function hasUnpairedEscapedSurrogate(source: string): boolean {
  const matches = [...source.matchAll(/\\u([0-9a-fA-F]{4})/g)];
  for (let index = 0; index < matches.length; index += 1) {
    const match = matches[index];
    const code = Number.parseInt(match?.[1] ?? "", 16);
    if (code >= 0xd800 && code <= 0xdbff) {
      const next = matches[index + 1];
      const nextCode = Number.parseInt(next?.[1] ?? "", 16);
      const adjacent = next?.index === (match?.index ?? 0) + (match?.[0].length ?? 0);
      if (!adjacent || nextCode < 0xdc00 || nextCode > 0xdfff) {
        return true;
      }
      index += 1;
    } else if (code >= 0xdc00 && code <= 0xdfff) {
      return true;
    }
  }
  return false;
}

function inspectNumber(
  node: Extract<ValueNode, { type: "Number" }>,
  source: string,
  pointer: string,
) {
  const token = source.slice(node.loc.start.offset, node.loc.end.offset);
  if (
    !Number.isFinite(node.value) ||
    Object.is(node.value, -0) ||
    /^-0(?:\.0*)?(?:[eE][+-]?0+)?$/.test(token)
  ) {
    throw new ContractInputError(
      "number",
      pointer,
      "Non-finite and negative-zero numbers are forbidden.",
    );
  }
  if (/^-?[0-9]+$/.test(token) && !Number.isSafeInteger(node.value)) {
    throw new ContractInputError(
      "number",
      pointer,
      "Integers outside the interoperable range must be encoded as strings.",
    );
  }
}

function inspectObject(node: ObjectNode, source: string, pointer: string): void {
  const names = new Set<string>();
  for (const member of node.members) {
    const name = member.name.type === "String" ? member.name.value : member.name.name;
    const memberPointer = childPointer(pointer, name);
    if (names.has(name)) {
      throw new ContractInputError("duplicate_key", memberPointer, `Duplicate object key: ${name}`);
    }
    names.add(name);
    const rawName =
      member.name.type === "String"
        ? source.slice(member.name.loc.start.offset, member.name.loc.end.offset)
        : name;
    if (hasUnpairedSurrogate(name) || hasUnpairedEscapedSurrogate(rawName)) {
      throw new ContractInputError(
        "unicode",
        pointer,
        "Object key contains an unpaired surrogate.",
      );
    }
    inspectNode(member.value, source, memberPointer);
  }
}

function inspectNode(node: ValueNode, source: string, pointer: string): void {
  switch (node.type) {
    case "Array":
      node.elements.forEach((element, index) => {
        inspectNode(element.value, source, childPointer(pointer, index));
      });
      return;
    case "Object":
      inspectObject(node, source, pointer);
      return;
    case "String":
      if (
        hasUnpairedSurrogate(node.value) ||
        hasUnpairedEscapedSurrogate(source.slice(node.loc.start.offset, node.loc.end.offset))
      ) {
        throw new ContractInputError("unicode", pointer, "String contains an unpaired surrogate.");
      }
      return;
    case "Number":
      inspectNumber(node, source, pointer);
      return;
    case "NaN":
    case "Infinity":
      throw new ContractInputError("number", pointer, "Non-finite numbers are forbidden.");
    case "Boolean":
    case "Null":
      return;
  }
}

function pointerForNonFiniteSyntax(source: string): string {
  return /"value"\s*:\s*(?:NaN|[-+]?Infinity)/.test(source) ? "/value" : "";
}

export function parseJsonStrict(source: string): JSONValue {
  let document;
  try {
    document = parse(source, { mode: "json", allowTrailingCommas: false });
  } catch (error) {
    if (/\b(?:NaN|Infinity)\b/.test(source)) {
      throw new ContractInputError(
        "number",
        pointerForNonFiniteSyntax(source),
        "Non-finite numbers are forbidden.",
      );
    }
    const message = error instanceof Error ? error.message : "Invalid JSON.";
    throw new ContractInputError("canonical", "", message);
  }

  inspectNode(document.body, source, "");
  return evaluate(document);
}

function assertCanonicalValue(
  value: unknown,
  pointer: string,
  seen: Set<object>,
): asserts value is JSONValue {
  if (value === null || typeof value === "boolean") {
    return;
  }
  if (typeof value === "string") {
    if (hasUnpairedSurrogate(value)) {
      throw new ContractInputError("unicode", pointer, "String contains an unpaired surrogate.");
    }
    return;
  }
  if (typeof value === "number") {
    if (
      !Number.isFinite(value) ||
      Object.is(value, -0) ||
      (Number.isInteger(value) && !Number.isSafeInteger(value))
    ) {
      throw new ContractInputError(
        "number",
        pointer,
        "Number is outside the canonical interoperable subset.",
      );
    }
    return;
  }
  if (typeof value !== "object") {
    throw new ContractInputError("type", pointer, "Value is not representable as JSON.");
  }
  if (seen.has(value)) {
    throw new ContractInputError("canonical", pointer, "Cyclic values cannot be canonicalized.");
  }
  seen.add(value);
  if (Array.isArray(value)) {
    value.forEach((item, index) => {
      assertCanonicalValue(item, childPointer(pointer, index), seen);
    });
  } else {
    for (const [key, item] of Object.entries(value)) {
      if (hasUnpairedSurrogate(key)) {
        throw new ContractInputError(
          "unicode",
          childPointer(pointer, key),
          "Object key contains an unpaired surrogate.",
        );
      }
      assertCanonicalValue(item, childPointer(pointer, key), seen);
    }
  }
  seen.delete(value);
}

export function canonicalJsonBytes(value: unknown): Buffer {
  assertCanonicalValue(value, "", new Set());
  const canonical = canonicalize(value);
  if (canonical === undefined) {
    throw new ContractInputError("canonical", "", "Value cannot be serialized as canonical JSON.");
  }
  return Buffer.from(canonical, "utf8");
}

export { childPointer, escapePointer };

File: python/src/palimpsest/contracts/canonical_json.py
from __future__ import annotations

import json
import math
from dataclasses import dataclass
from typing import Any

import rfc8785


@dataclass(frozen=True)
class ContractInputError(ValueError):
    reason: str
    pointer: str
    message: str

    def __str__(self) -> str:
        return self.message


@dataclass(frozen=True)
class _NumberToken:
    raw: str


class _ObjectPairs(list[tuple[str, Any]]):
    pass


def _escape_pointer(value: str) -> str:
    return value.replace("~", "~0").replace("/", "~1")


def child_pointer(parent: str, child: str | int) -> str:
    return f"{parent}/{child if isinstance(child, int) else _escape_pointer(child)}"


def _has_surrogate(value: str) -> bool:
    return any(0xD800 <= ord(character) <= 0xDFFF for character in value)


def _convert_number(token: _NumberToken, pointer: str) -> int | float:
    raw = token.raw
    if any(character in raw for character in ".eE"):
        value = float(raw)
        if not math.isfinite(value) or (value == 0 and raw.startswith("-")):
            raise ContractInputError(
                "number",
                pointer,
                "Non-finite and negative-zero numbers are forbidden.",
            )
        return value

    value = int(raw)
    if raw == "-0":
        raise ContractInputError("number", pointer, "Negative zero is forbidden.")
    if abs(value) > 9_007_199_254_740_991:
        raise ContractInputError(
            "number",
            pointer,
            "Integers outside the interoperable range must be encoded as strings.",
        )
    return value


def _convert(value: Any, pointer: str) -> Any:
    if isinstance(value, _ObjectPairs):
        result: dict[str, Any] = {}
        for name, child in value:
            member_pointer = child_pointer(pointer, name)
            if name in result:
                raise ContractInputError(
                    "duplicate_key",
                    member_pointer,
                    f"Duplicate object key: {name}",
                )
            if _has_surrogate(name):
                raise ContractInputError(
                    "unicode",
                    pointer,
                    "Object key contains a surrogate.",
                )
            result[name] = _convert(child, member_pointer)
        return result
    if isinstance(value, list):
        return [_convert(child, child_pointer(pointer, index)) for index, child in enumerate(value)]
    if isinstance(value, _NumberToken):
        return _convert_number(value, pointer)
    if isinstance(value, str) and _has_surrogate(value):
        raise ContractInputError("unicode", pointer, "String contains an unpaired surrogate.")
    return value


def _reject_constant(raw: str) -> None:
    raise ContractInputError("number", "/value", f"Non-finite number is forbidden: {raw}")


def parse_json_strict(source: str) -> Any:
    try:
        parsed = json.loads(
            source,
            object_pairs_hook=_ObjectPairs,
            parse_constant=_reject_constant,
            parse_float=_NumberToken,
            parse_int=_NumberToken,
        )
    except ContractInputError:
        raise
    except json.JSONDecodeError as error:
        raise ContractInputError("canonical", "", str(error)) from error
    return _convert(parsed, "")


def _assert_canonical_value(value: Any, pointer: str, seen: set[int]) -> None:
    if value is None or isinstance(value, bool):
        return
    if isinstance(value, str):
        if _has_surrogate(value):
            raise ContractInputError("unicode", pointer, "String contains an unpaired surrogate.")
        return
    if isinstance(value, int):
        if abs(value) > 9_007_199_254_740_991:
            raise ContractInputError(
                "number",
                pointer,
                "Integer is outside the interoperable range.",
            )
        return
    if isinstance(value, float):
        if not math.isfinite(value) or (value == 0 and math.copysign(1, value) < 0):
            raise ContractInputError("number", pointer, "Number is outside the canonical subset.")
        return
    if not isinstance(value, (dict, list)):
        raise ContractInputError("type", pointer, "Value is not representable as JSON.")
    identity = id(value)
    if identity in seen:
        raise ContractInputError("canonical", pointer, "Cyclic values cannot be canonicalized.")
    seen.add(identity)
    if isinstance(value, list):
        for index, child in enumerate(value):
            _assert_canonical_value(child, child_pointer(pointer, index), seen)
    else:
        for name, child in value.items():
            if not isinstance(name, str):
                raise ContractInputError("type", pointer, "JSON object names must be strings.")
            _assert_canonical_value(child, child_pointer(pointer, name), seen)
    seen.remove(identity)


def canonical_json_bytes(value: Any) -> bytes:
    _assert_canonical_value(value, "", set())
    try:
        return rfc8785.dumps(value)
    except (rfc8785.CanonicalizationError, UnicodeEncodeError) as error:
        raise ContractInputError("canonical", "", str(error)) from error
